device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/auth/IotHubSasTokenHardwareAuthenticationProvider.java

```diff
@@ -65,14 +65,19 @@ public IotHubSasTokenHardwareAuthenticationProvider(String hostname, String gate
     /**
      * Getter for SasToken. If the saved token has expired, this method shall renew it if possible
      *
+     * @param proactivelyRenew if true, this method will generate a fresh sas token even if the previously saved token
+     *                                 has not expired yet as long as the current token has lived beyond its buffer.
+     *                                 Use this for pre-emptively renewing sas tokens.
+
      * @throws IOException if generating the sas token from the TPM fails
      * @return The value of SasToken
      */
-    public String getRenewedSasToken() throws IOException
+    public String getRenewedSasToken(boolean proactivelyRenew) throws IOException
     {
-        if (this.sasToken.isExpired())
+        if (this.sasToken.isExpired() || (proactivelyRenew && this.shouldRefreshToken()))
```

Comment:
isExpired() is check within shouldRefreshToken().  Maybe we could remove it from here...
I would prefer calling "shouldRefreshToken" alone in the All AuthenticationProviders to make the logic cleaner...  maybe passing proactivelyRenew inside shouldRefreshToken??

Reply:
I'll see if I can implement it that way instead. I see what you are getting at.